### DISCUSSION TITLE
Different? Smart Socket

### DISCUSSION THREAD

**Discussion by jamesonuk** [make-all/tuya-local#1160] [Q&A]
I have what is marketed in the UK as a Knightsbridge Dual outdoor socket.

When I add this I get several different devices to choose from but none seem to exactly match the dps listed against the device.
I am using dual_power_monitor_smartplugv2.yaml which is one of the options and seems to map most things OK.

The one thing I really want is the power consumption from the device (not current power).   DP 17 is `add_ele` which I guess is the kWh data I want *and shows up in the Tuya app?    This isn't showing up anywhere but I couldn't quite work out he yaml definitions.

The definition I am using https://github.com/make-all/tuya-local/blob/main/custom_components/tuya_local/devices/dual_power_monitor_smartplugv2.yaml has dp 17 listed under the primary entity but marked as optional where as one of the other definitions I was looking at had it under secondary entities (as the current / voltage / power sensors are) but not sure what the difference is.

So the main question is whether the existing definition is fine for this device or whether I need to adjust or create a new one?
Is the `add_ele `value mapped properly and the fact it is not showing up because although the device says it is one of the DPs it is not actually being sent?   Is there something else I need to do to get the power consumption figures that show up in the app (I would prefer to use that rather than the Riemann Sum Integral I am using now based on the power

For reference the list of options for the device is

- digoo_dgsp202
- dual_power_monitor_smartplugv2
- mellerware_citymove_vacuum
- pj1103_clamp_meter
- simple_dual_switch_timer_v2
- simple_switch
- simple_swich_timerv2
- smartplugv2
- smartplugv2_energy
- smartplugv2_polled_power

Although several of these appear to be single sockets but the tip two both work (albeit with a few different DPs)

Device details are
```
    "model": "SN9KW/CU9KW/OP9KW",
    "name": "Socket",
    "product_id": "mgewa9czeq3byvj3",
    "product_name": "Socket",
    "sub": false,
```
and query things data model gives
```
{    
	"modelId": "0000003bgs",
    "services": [
        {
            "actions": [],
            "code": "",
            "description": "",
            "events": [],
            "name": "默认服务",
            "properties": [
                {
                    "abilityId": 1,
                    "accessMode": "rw",
                    "code": "switch_1",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_power2",
                        "attribute": "16"
                    },
                    "name": "开关1",
                    "typeSpec": {
                        "type": "bool",
                        "typeDefaultValue": false
                    }
                },
                {
                    "abilityId": 2,
                    "accessMode": "rw",
                    "code": "switch_2",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_power2",
                        "attribute": "16"
                    },
                    "name": "开关2",
                    "typeSpec": {
                        "type": "bool",
                        "typeDefaultValue": false
                    }
                },
                {
                    "abilityId": 9,
                    "accessMode": "rw",
                    "code": "countdown_1",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_time2"
                    },
                    "name": "开关1倒计时",
                    "typeSpec": {
                        "max": 86400,
                        "min": 0,
                        "scale": 0,
                        "step": 1,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": "s"
                    }
                },
                {
                    "abilityId": 10,
                    "accessMode": "rw",
                    "code": "countdown_2",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_time2"
                    },
                    "name": "开关2倒计时",
                    "typeSpec": {
                        "max": 86400,
                        "min": 0,
                        "scale": 0,
                        "step": 1,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": "s"
                    }
                },
                {
                    "abilityId": 17,
                    "accessMode": "rw",
                    "code": "add_ele",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-battery"
                    },
                    "name": "增加电量",
                    "typeSpec": {
                        "max": 50000,
                        "min": 0,
                        "scale": 3,
                        "step": 100,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": ""
                    }
                },
                {
                    "abilityId": 18,
                    "accessMode": "ro",
                    "code": "cur_current",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-Ele"
                    },
                    "name": "当前电流",
                    "typeSpec": {
                        "max": 30000,
                        "min": 0,
                        "scale": 0,
                        "step": 1,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": "mA"
                    }
                },
                {
                    "abilityId": 19,
                    "accessMode": "ro",
                    "code": "cur_power",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_tool"
                    },
                    "name": "当前功率",
                    "typeSpec": {
                        "max": 50000,
                        "min": 0,
                        "scale": 1,
                        "step": 1,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": "W"
                    }
                },
                {
                    "abilityId": 20,
                    "accessMode": "ro",
                    "code": "cur_voltage",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-a_function_turbo"
                    },
                    "name": "当前电压",
                    "typeSpec": {
                        "max": 5000,
                        "min": 0,
                        "scale": 1,
                        "step": 1,
                        "type": "value",
                        "typeDefaultValue": 0,
                        "unit": "V"
                    }
                }
            ]
        }
    ]
}
```

**Comment by make-all** [comment]
DP 17 rarely contains usable data in Tuya smartplugs.  They seem to reset the counter as soon as the data is reported to the cloud.  You should just use the power sensor with an integration helper to track energy usage.

**Comment by jamesonuk** [comment]
> DP 17 rarely contains usable data in Tuya smartplugs. They seem to reset the counter as soon as the data is reported to the cloud. You should just use the power sensor with an integration helper to track energy usage.

😢 which is kind of what I figured but it is really annoying.   Because the power jumps between extremes the integral sensor isn't that accurate.

Is this data polled or event based?   I am guessing that DP17 is sending an amount to add to the total each time as an event so will generally be zero and will then fire off an event once some threshold is met.   Just wondering whether whatever is sent on DP17 would show up or not (mainly out of interest to try and see if there is anything I could salvage from it)

**Comment by jamesonuk** [comment]
OK I have done some more digging and the device is reporting on DP id 17 every so often (seems to update quickly when power is being used or at least every half hour if not).
![image](https://github.com/make-all/tuya-local/assets/1040621/2810d4c5-bba9-4671-9ce6-da8c7aa1002f)

Summing these values for the day gives the value you see in the app and it can be quite different to what is estimated with the Riemann Sum Integral.   Each value is the amount of energy used since the last report rather than a measurement but I think it is useful.  If I can get the values into a sensor in HA I believe I can use a Utility Meter helper on top that is set for the values to be deltas and it will handle the rest.

I updated the device definition to include
```
  - entity: sensor
    name: Energy Increase
    class: energy
    dps:
      - id: 17
        type: integer
        name: sensor
        unit: kWh
```
but after the device reporting the sensor didn't update
![image](https://github.com/make-all/tuya-local/assets/1040621/b9f16045-b581-4eca-97f7-6c24b1ba9919)

Have I missed something?  Is my definition wrong or is there something that would prevent the values getting from the device to HA (is this polling device for attributes rather than listening for pushed events?)

**Reply by make-all** [reply]
Your definition is not wrong, the plug is not reporting energy locally, only to the cloud
